src/Chapter3Task.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE InstanceSigs #-}
+
+module Chapter3Task where
+
+newtype Health = MkHealth {unHealth :: Integer} deriving (Show, Eq)
+
+newtype Attack = MkAttack {unAttack :: Integer} deriving (Show, Eq)
+
+newtype Defense = MkDefense {unDefense :: Integer} deriving (Show, Eq)
+
+data KnightAction
+  = KnightAttack 
+  | DrinkPotion 
+  | CastSpell 
+  deriving (Show, Eq)
+
+data MonsterAction
+  = MonsterAttack
+  | RunAway
+  deriving (Show, Eq)
+
+data Action
+  = KnightAction KnightAction
+  | MonsterAction MonsterAction
+  deriving (Show, Eq)
+
+data Knight = Knight {
+  knightHealth :: Health,
+  knightAttack :: Attack,
+  knightDefense :: Defense,
+  knightActions :: [Action] -- is there a way to limit to KnightActions only here? Can this lead to runtime errors?
+} deriving (Show, Eq)
+
+data Monster = Monster {
+  monsterHealth :: Health,
+  monsterAttack :: Attack,
+  monsterActions :: [Action]
+} deriving (Show, Eq)
+
+class Player a where
+  getHealth :: a -> Health
+  getAttack :: a -> Attack
+  getActions :: a -> [Action]
+  getDefense :: a -> Defense
+  setHealth :: a -> Health -> a
+  setDefense :: a -> Defense -> a
+  setActions :: a -> [Action] -> a
+
+instance Player Knight where
+    getHealth = knightHealth
+    getAttack = knightAttack
+    getActions = knightActions
+    getDefense = knightDefense
+    setHealth knight health = knight {knightHealth = health}
+    setDefense knight defense = knight {knightDefense = defense}
+    setActions knight actions = knight {knightActions = actions}
+
+instance Player Monster where
+    getHealth = monsterHealth
+    getAttack = monsterAttack
+    getActions = monsterActions
+    getDefense _ = MkDefense 0
+    setHealth monster _ = monster
+    setDefense monster _ = monster -- should this be allowed?
```

Comment:
You can design typeclasses differently to avoid the need to have the `getActions` method and methods like `setDefense`, so there won't be a possibility to set defense for a character who doesn't have the defense.